engines/registry/app/models/registry/authenticated_client.rb

```diff
@@ -16,7 +16,7 @@ def initialize(login, password)
 
   def authenticate_by_system_credentials(login, password)
     @systems = System.get_by_credentials(login, password)
-    if @systems.present?
+    if @systems.any? { |system| system.last_seen_at > Settings[:registry].token_expiration.seconds.ago }
```

Comment:
The `last_seen_at` is touched on each API interaction of the client, so by querying the API each hour, the client would always get authenticated here. Not sure if that's an issue though.

Reply:
Yes it would be. The idea is to expire the token after 8 hours no matter how often the system comes to visit in those 8 hours.

Reply:
As a user, if auth is valid for 8 hours, and I re-auth every, say 4 hours, I should never encounter a time when I cannot access the system. Forcing expiration, even with a re-auth, doesn't make sense to me.

Reply:
I see the behavior correct as it is, if the re-auth happens after X amount of time, default 8h, is because the credentials have not been refreshed/used, meaning there is no guarantee that after that time the system is still registered or valid or whatever state the system is

If during those 8h, system calls any API that updates `last_seen_at` that guarantees that the system is registered and can have access to stuff that required registration/auth, like the registry

Reply:
I think we need to pick this apart some more to make sure we are all on the same page.

"auth" i.e. verification of the system credentials against the IMDS data happens only in 2 ways:

1. The RMT server is accessed with zypper, because zypper always sends the IMDS data and if the system is not in the cache then the full verification path takes place
2. The user executes a forced re-auth via the new command we will provide client side.

If the update of `last_seen_at` in the DB only happens when zypper comes to visit then the implementation as is in this PR works as expected because we are guaranteed that the proper auth happened. Repository access expires after 20 minutes and then a full verification happens when the repository path is visited again. However, if `last_seen_at` is updated on the registry path as well then the implementation is not correct because we cannot do a full verification on the registry path since the client, docker or podman, have no way of sending the metadata.

My question is, when is `last_seen_at` updated? 
@digitaltom states "on each API interaction of the client" which would be every time the heartbeat beacon runs, i.e. hourly, and then we are not meeting the requirement of "a full verification every 8 hours for the registry path." It could also be when the registry path is accessed, same problem.

We would only meet the requirement if `last_seen_at` is updated **only** on the repository access path. According to the comment from @digitaltom this is not true. Therefore the implementation in this PR does not meet the requirement.

It appears to me, based on [1] that `last_seen_at` is updated only on successful authentication of a system (instance) which in the Public Cloud means access via the repository path. Somone with more knowledge about the details will need to confirm this please.

[1] https://github.com/SUSE/rmt/blob/master/app/controllers/application_controller.rb#L25

Reply:
> My question is, when is last_seen_at updated?

It gets updated with `zypper` calls (e.g.`zypper ref`, `zypper up`, `zypper in <whatever>`) if done at least 3min apart [2] or `SUSEConnect -p <whatever>` 
Meaning if the system's `last_seen_at` is at, let's say 10:04:00, and expiration is set to be at 10:05:00 and user ran any of those commands,`last_seen_at` would change to, let's say 10:07:00 then user would not need to run reauth command as `zypper` updated that time and access would be granted which I understood to be OK behavior

> The idea is to expire the token after 8 hours no matter how often the system comes to visit in those 8 hours

I understood the behavior of the PR correct until that comment, if we want the credentials to expire after 8 hours, independently of how many times `zypper` or `SUSEConnect -p` are called (which is what that comment states), or how many times `last_seen_at` has been updated (if any) then we'd need another way. If we do not mind that the system access for the registry has been granted by running `zypper` and that it may not expire every 8 hours because of the same reason, then this PR is OK 

IIRC, `last_seen_at` in the public cloud gets updated on [2], as in the public cloud the system token is disabled


[2] https://github.com/SUSE/rmt/blob/master/app/controllers/application_controller.rb#L30

Reply:
My comment 

> The idea is to expire the token after 8 hours no matter how often the system comes to visit in those 8 hours

is confusing, you are correct. I should have been more explicit it should be:

**The token expires after 8 hours no matter how often the client comes to visit on the registry path in those 8 hours.**

With this, since a full validation has to happen before `last_seen_at` gets updated this implementation will work. Thanks for the clarification.